### PR TITLE
Add Preseq

### PR DESCRIPTION
**Pull request by jbv2**
Adds `test_humanbam` test profile. 
Adds preseq.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [x] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jbv2** [comment]
@nf-core-bot fix linting

File: conf/modules.config [hunk 1]
```diff
@@ -34,6 +34,7 @@ process {
         ]
     }
 
+
```
Comment (jfy133) [inline conf/modules.config:37]:
```suggestion
```

File: docs/output.md [hunk 2]
```diff
@@ -176,3 +176,16 @@ Deduplication is carried by two possible tools, as described below. However the
...
 [DeDup](https://github.com/apeltzer/DeDup) is a merged-read deduplication tool capable of performing merged-read deduplication on paired-end sequencing data in BAM files.
+
+#### Preseq
+
+<details markdown="1">
+<summary>Output files</summary>
+
+- `mapping/`
+
+  - `*.c_curve.txt`: A file for every BAM file that had enough deduplication statistics to generate a complexity curve for estimating the amount unique reads that will be yield if the library is re-sequenced. You can use this file for plotting e.g. in R to find your sequencing target depth.
```
Comment (jfy133) [inline docs/output.md:187]:
```suggestion
  - `*.c_curve.txt`: A two column text file with the first column representing sequencing depth and the second an estimate of unique reads .
```

File: docs/output.md [hunk 3]
```diff
@@ -176,3 +176,16 @@ Deduplication is carried by two possible tools, as described below. However the
...
+
+  </details>
+
+  [Preseq](https://github.com/smithlabcode/preseq) Preseq is a collection of tools that allow assessment of the complexity of the library, where complexity means the number of unique molecules in your library (i.e. not molecules with the exact same length and sequence). There are two algorithms from the tools we use: c_curve and lc_extrap. The former gives you the expected number of unique reads if you were to repeated sequencing but with fewer reads than your first sequencing run. The latter tries to extrapolate the decay in the number of unique reads you would get with re-sequencing but with more reads than your initial sequencing run.
```
Comment (jfy133) [inline docs/output.md:191]:
```suggestion
  [Preseq](https://github.com/smithlabcode/preseq) Preseq is a collection of tools that allow assessment of the complexity of the library, where complexity means the number of unique molecules in your library (i.e. not molecules with the exact same length and sequence). There are two algorithms from the tools we use: c_curve and lc_extrap. The former gives you the expected number of unique reads if you were to repeated sequencing but with fewer reads than your first sequencing run. The latter tries to extrapolate the decay in the number of unique reads you would get with re-sequencing but with more reads than your initial sequencing run.
  
The resulting histogram file will contain estimated deduplication statistics at different theoretical sequencing depths, and can be used to generate a complexity curve for estimating the amount unique reads that will be yield if the library is re-sequenced. 

These curves will be displayed in the pipeline run's MultiQC report, however you can also use this file for plotting yourself for further exploration e.g. in R to find your sequencing target depth.
```

File: nextflow_schema.json [hunk 4]
```diff
@@ -571,14 +571,29 @@
...
-                    "help_text": "Sets the duplicate read removal tool. By default uses `markduplicates` from Picard. Alternatively an ancient DNA specific read deduplication tool `dedup` (Peltzer et al. 2016) is offered. The latter utilises both ends of paired-end data to remove duplicates (i.e. true exact duplicates, as markduplicates will over-zealously deduplicate anything with the same starting position even if the ends are different).\n\n> ⚠️ DeDup can only be used on collapsed (i.e. merged) reads from paired-end sequencing.",
-                    "enum": ["markduplicates", "dedup"]
+                    "help_text": "Sets the duplicate read removal tool. By default uses `markduplicates` from Picard. Alternatively an ancient DNA specific read deduplication tool `dedup` (Peltzer et al. 2016) is offered. The latter utilises both ends of paired-end data to remove duplicates (i.e. true exact duplicates, as markduplicates will over-zealously deduplicate anything with the same starting position even if the ends are different).\n\n> \u26a0\ufe0f DeDup can only be used on collapsed (i.e. merged) reads from paired-end sequencing.",
```
Comment (jfy133) [inline nextflow_schema.json:581]:
```suggestion
                    "help_text": "Sets the duplicate read removal tool. By default uses `markduplicates` from Picard. Alternatively an ancient DNA specific read deduplication tool `dedup` (Peltzer et al. 2016) is offered. The latter utilises both ends of paired-end data to remove duplicates (i.e. true exact duplicates, as markduplicates will over-zealously deduplicate anything with the same starting position even if the ends are different).\n\n> ⚠️ DeDup can only be used on collapsed (i.e. merged) reads from paired-end sequencing.",
```

File: nextflow_schema.json [hunk 5]
```diff
@@ -571,14 +571,29 @@
...
+                "deduplication_skip_preseq": {
+                    "type": "boolean",
+                    "help_text": "Turns off the computation of library complexity estimation.",
+                    "description": "Turns off the computation of library complexity estimation.",
+                    "fa_icon": "fas fa-forward"
+                },
+                "deduplication_preseq_stepsize": {
+                    "type": "integer",
+                    "default": 1000,
+                    "description": "Specify the step size of Preseq.",
```
Comment (jfy133) [inline nextflow_schema.json:594]:
```suggestion
                    "description": "Specify the step size (i.e., sampling regularity) of Preseq.",
```

File: nextflow_schema.json [hunk 6]
```diff
@@ -571,14 +571,29 @@
...
+                    "help_text": "Can be used to configure the step size of Preseq's c_curve and lc_extrap method. Can be useful when only few and thus shallow sequencing results are used for extrapolation.\n\n\n> Modifies tool parameter(s)\n>   - preseq: `-s`",
```
Comment (jfy133) [inline nextflow_schema.json:595]:
```suggestion
                    "help_text": "Can be used to configure the step size of Preseq's c_curve and lc_extrap method. Can be useful when few reads and allow Preseq to be used for extrapolation of shallow sequencing results.\n\n\n> Modifies tool parameter(s)\n>   - preseq: `-s`",
```

File: workflows/eager.nf [hunk 7]
```diff
@@ -198,16 +199,24 @@ workflow EAGER {
...
-
-        ch_dedupped_bams     = DEDUPLICATE.out.bam.join( DEDUPLICATE.out.bai )
+        ch_dedupped_bams = DEDUPLICATE.out.bam
+            .join( DEDUPLICATE.out.bai )
         ch_dedupped_flagstat = DEDUPLICATE.out.flagstat
+        ch_versions = ch_versions.mix( DEDUPLICATE.out.versions )
 
     } else {
-        ch_dedupped_bams     = ch_reads_for_deduplication
+        ch_dedupped_bams = ch_reads_for_deduplication
```
Comment (jfy133) [inline workflows/eager.nf:208]:
```suggestion
        ch_dedupped_bams     = ch_reads_for_deduplication
```

File: workflows/eager.nf [hunk 8]
```diff
@@ -226,7 +235,7 @@ workflow EAGER {
     ch_multiqc_files = ch_multiqc_files.mix(ch_workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))
     ch_multiqc_files = ch_multiqc_files.mix(ch_methods_description.collectFile(name: 'methods_description_mqc.yaml'))
     ch_multiqc_files = ch_multiqc_files.mix(CUSTOM_DUMPSOFTWAREVERSIONS.out.mqc_yml.collect())
-    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]}.ifEmpty([]))
+
```
Comment (jfy133) [inline workflows/eager.nf:238]:
```suggestion
    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]}.ifEmpty([]))
```

Comment (jfy133) [reply workflows/eager.nf:238]: We still need this! Not sure If I said the wrong thing during teh session - sorry!

**Review by jfy133 — COMMENTED** [review]
Overall code looks good :) - small tweaks here and there

**Review by jfy133 — CHANGES_REQUESTED** [review]
We should add the LCextrap module as well! I forgot we added that to eager2, and we refer to it in the output docs!

It should be quite straight forward, you can partly copy Selina's if statmeetn for Falco/FastQC in #949

File: workflows/eager.nf [hunk 9]
```diff
@@ -205,16 +207,30 @@ workflow EAGER {
 
     if ( !params.skip_deduplication ) {
         DEDUPLICATE( ch_reads_for_deduplication, ch_fasta_for_deduplication.fasta, ch_fasta_for_deduplication.fasta_fai )
-        ch_versions          = ch_versions.mix( DEDUPLICATE.out.versions )
-
-        ch_dedupped_bams     = DEDUPLICATE.out.bam.join( DEDUPLICATE.out.bai )
+        ch_dedupped_bams = DEDUPLICATE.out.bam
+            .join( DEDUPLICATE.out.bai )
         ch_dedupped_flagstat = DEDUPLICATE.out.flagstat
+        ch_versions = ch_versions.mix( DEDUPLICATE.out.versions )
```
Comment (TCLamnidis) [inline workflows/eager.nf:213]:
```suggestion
        ch_dedupped_flagstat = DEDUPLICATE.out.flagstat
        ch_versions                   = ch_versions.mix( DEDUPLICATE.out.versions )
```

File: workflows/eager.nf [hunk 10]
```diff
@@ -205,16 +207,30 @@ workflow EAGER {
...
 
     } else {
         ch_dedupped_bams     = ch_reads_for_deduplication
         ch_dedupped_flagstat = Channel.empty()
     }
 
+    //
+    // MODULE: PreSeq
+    //
+    if ( !params.mapstats_skip_preseq && params.mapstats_preseq_mode == 'c_curve') {
```
Comment (jfy133) [inline workflows/eager.nf:223]: Why two separate `if` statements? Can be combined into one `if/else`

File: workflows/eager.nf [hunk 11]
```diff
@@ -233,6 +249,7 @@ workflow EAGER {
     ch_multiqc_files = ch_multiqc_files.mix(ch_workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))
     ch_multiqc_files = ch_multiqc_files.mix(ch_methods_description.collectFile(name: 'methods_description_mqc.yaml'))
     ch_multiqc_files = ch_multiqc_files.mix(CUSTOM_DUMPSOFTWAREVERSIONS.out.mqc_yml.collect())
+    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]}.ifEmpty([]))
```
Comment (jfy133) [inline workflows/eager.nf:252]: Isn't this defined further up?

**Review by jfy133 — COMMENTED** [review]
Mostly minor comments, but otherwise LGTM

File: workflows/eager.nf [hunk 12]
```diff
@@ -205,16 +207,30 @@ workflow EAGER {
...
 
+    //
+    // MODULE: PreSeq
+    //
+    if ( !params.mapstats_skip_preseq && params.mapstats_preseq_mode == 'c_curve') {
+        PRESEQ_CCURVE(ch_reads_for_deduplication.map{[it[0],it[1]]})
+        ch_multiqc_files = ch_multiqc_files.mix(PRESEQ_CCURVE.out.c_curve.collect{it[1]}.ifEmpty([]))
+        ch_versions = ch_versions.mix( PRESEQ_CCURVE.out.versions )
+    }
+
```
Comment (jbv2) [inline workflows/eager.nf:228]:
```suggestion
} else 
```

File: workflows/eager.nf (hunk 10 above)
Comment (jbv2) [reply workflows/eager.nf:223]: Agree

Comment (jbv2) [reply workflows/eager.nf:223]: Agree

File: workflows/eager.nf (hunk 11 above)
Comment (jbv2) [reply workflows/eager.nf:252]: Yes, can it just be deleted?

**Review by jbv2 — COMMENTED** [review]
Accepting suggestions

**Review by jbv2 — COMMENTED** [review]
Accepting suggestions

File: workflows/eager.nf (hunk 11 above)
Comment (jfy133) [reply workflows/eager.nf:250]: Yes!

File: workflows/eager.nf [hunk 13]
```diff
@@ -233,6 +247,7 @@ workflow EAGER {
     ch_multiqc_files = ch_multiqc_files.mix(ch_workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))
     ch_multiqc_files = ch_multiqc_files.mix(ch_methods_description.collectFile(name: 'methods_description_mqc.yaml'))
     ch_multiqc_files = ch_multiqc_files.mix(CUSTOM_DUMPSOFTWAREVERSIONS.out.mqc_yml.collect())
+    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]}.ifEmpty([]))
```
Comment (jbv2) [inline workflows/eager.nf:250]:
```suggestion
```

**Review by jbv2 — COMMENTED** [review]
Removing duplicated

File: nextflow.config [hunk 14]
```diff
@@ -203,8 +213,9 @@ profiles {
         executor.cpus          = 16
         executor.memory        = 60.GB
     }
-    test      { includeConfig 'conf/test.config'      }
-    test_full { includeConfig 'conf/test_full.config' }
+    test            { includeConfig 'conf/test.config'          }
+    test_full       { includeConfig 'conf/test_full.config'
+    test_humanbam   { includeConfig 'conf/test_humanbam.config' }
```
[the remark this answers is not in the document]
Comment (jfy133) [reply nextflow.config:218]: My mistake before :grimacing: sorry

File: workflows/eager.nf (hunk 13 above)
Comment (jfy133) [inline workflows/eager.nf:250]:
```suggestion
```

**Review by jfy133 — APPROVED** [review]

**Review by jfy133 — APPROVED** [review]
@jbv2 once tests pass you can merge :)

**Pull request merged**